### PR TITLE
beatrix: Add invoices tests to verify behavior when doing full item a…

### PR DESCRIPTION
**Pull request by sbrossie**
…djustements

### CODE REVIEW

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java [hunk 1]
```diff
@@ -890,8 +823,137 @@ public void testWithSuperflousRepairedItems() throws Exception {
                 new ExpectedInvoiceItemCheck(new LocalDate(2012, 6, 1), new LocalDate(2012, 7, 1), InvoiceItemType.RECURRING, new BigDecimal("249.95")));
         invoiceChecker.checkInvoice(invoices.get(2).getId(), callContext, toBeChecked);
 
+    }
+
+    //
+    //    Full item adjustments scenarios
+    //
+    //    Case V1: Cancellation EOT => Balance is $0 and nothing to regenerate
```
Comment (pierre) [inline beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java]: `SOT`, not `EOT`?

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java [hunk 2]
```diff
@@ -890,8 +823,137 @@ public void testWithSuperflousRepairedItems() throws Exception {
...
+
+        // Move clock to 2013-09-17
+        clock.addDays(30);
+        busHandler.pushExpectedEvents(NextEvent.BLOCK, NextEvent.CANCEL, NextEvent.NULL_INVOICE);
+        bpEntitlement.cancelEntitlementWithPolicyOverrideBillingPolicy(EntitlementActionPolicy.IMMEDIATE, BillingActionPolicy.START_OF_TERM, ImmutableList.<PluginProperty>of(), callContext);
+        assertListenerStatus();
+
+        List<Invoice>  invoices = invoiceUserApi.getInvoicesByAccount(bpEntitlement.getAccountId(), false, false, callContext);
+        assertEquals(invoices.size(), 2);
+        ImmutableList<ExpectedInvoiceItemCheck> toBeChecked = ImmutableList.<ExpectedInvoiceItemCheck>of(
```
Comment (pierre) [inline beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java]: Nit: could do the check before moving the clock to show that these items already existed and nothing was generated.

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java [hunk 3]
```diff
@@ -890,8 +823,137 @@ public void testWithSuperflousRepairedItems() throws Exception {
...
+        // FULL ITEM ADJUSTMENT PRIOR TO DOING THE REPAIR
+        //
+        final Invoice invoice1 = invoices.get(1);
+        final List<Payment> payments = paymentApi.getAccountPayments(account.getId(), false, false, ImmutableList.<PluginProperty>of(), callContext);
+        final Payment payment1 = payments.get(0);
+
+        final Map<UUID, BigDecimal> iias = new HashMap<UUID, BigDecimal>();
+        iias.put(invoice1.getInvoiceItems().get(0).getId(), null);
+        refundPaymentWithInvoiceItemAdjAndCheckForCompletion(account, payment1, iias, NextEvent.PAYMENT, NextEvent.INVOICE_PAYMENT, NextEvent.INVOICE_ADJUSTMENT);
+        checkNoMoreInvoiceToGenerate(account);
```
Comment (pierre) [inline beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java]: Could you trigger an invoice generation here to make sure it's a no-op?

Creating the adjustments bypasses the tree code, and before even triggering a subscription change, I'd like to make sure the invoice is stable.

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java (hunk 1 above)
Comment (sbrossie) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java]: Yep, will fix.

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java (hunk 2 above)
Comment (sbrossie) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java]: @pierre  Isn't it what [this line](https://github.com/killbill/killbill/blob/work-for-invoice-0.21.x/beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java#L954) is doing though?

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java (hunk 3 above)
Comment (sbrossie) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java]: @pierre  Isn't it what [this line](https://github.com/killbill/killbill/blob/work-for-invoice-0.21.x/beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java#L954) is doing though?

Comment (pierre) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationInvoiceWithRepairLogic.java]: Ah, yes, looks like it! 👍

**Pull request merged**